cpp/examples/protobuf/reader.cpp

```diff
@@ -0,0 +1,152 @@
+// Example code for reading Protobuf messages from an MCAP file. Uses the proto definition
+// from within the MCAP, with no dependency on generated headers.
+// Try it out by generating some PointCloud messages with the protobuf writer example,
+// and running this executable with the resulting MCAP file.
+#define MCAP_IMPLEMENTATION
+#include "mcap/reader.hpp"
+
+#include <google/protobuf/descriptor.pb.h>
+#include <google/protobuf/dynamic_message.h>
+
+#include <functional>
+#include <unordered_set>
+#include <vector>
+
+namespace gp = google::protobuf;
+
+bool WriteFileDescriptorsToPool(const gp::FileDescriptorSet* fdSet, gp::DescriptorPool* pool) {
+  // The FileDescriptorSet structure contains a repeated field of FileDescriptorProto instances.
+  // Adding these to the DescriptorPool should be as simple as calling pool->BuildFile,
+  // however BuildFile fails if any of a FileDescriptorProto's dependencies have not already been
+  // added.
+  // Therefore, do a topological sort first to put the FileDescriptorProtos in dependency order.
+  // we use the technique described here:
+  // https://en.wikipedia.org/wiki/Topological_sorting#Depth-first_search
```

Comment:
I wonder if it would work to Add() each FileDescriptorProto into a SimpleDescriptorDatabase? https://github.com/protocolbuffers/protobuf/blob/520c601c99012101c816b6ccc89e8d6fc28fdbb8/src/google/protobuf/descriptor_database.h#L144-L165